packages/x-components/src/components/dynamic-props.mixin.ts

```diff
@@ -0,0 +1,37 @@
+import Vue from 'vue';
+import { PropValidator } from 'vue/types/options';
+import { ExtendedVue } from 'vue/types/vue';
+
+/**
+ * Mixin that creates a string prop in the component for each element
+ * within the array passed as `propNames` argument.
+ *
+ * @param propNames - Array with the names of the props to create.
+ * @example
+ * ```typescript
+ *
+ *   const nodes = ['list', 'listitem', 'tag'] as const;
+ *
+ *   @Component({
+ *     components: { RelatedTag },
+ *     mixins: [xComponentMixin(relatedTagsXModule)]
+ *   })
+ *   export default class RelatedTags extends mixins(
+ *     dynamicPropsMixin<ExtractArrayItems<typeof nodes>>(nodes)
+ *   ) {
+ *   // This component will have available 3 props: 'list', 'listitem', 'tag'
+ *   }
+ * ```
+ * @returns Mixin for the component.
+ *
+ */
+export function dynamicPropsMixin<T extends string>(
+  propNames: readonly string[]
+): ExtendedVue<Vue, unknown, unknown, unknown, Record<T, string>> {
+  return Vue.extend({
+    props: propNames.reduce(
+      (a, v) => ({ ...a, [v]: { type: String } }),
+      {} as Record<T, PropValidator<string>>
+    )
+  });
+}
```

Comment:
A couple of suggestions after talking with you:

- Just by linking the generic parameter with the function parameter you will gain type inference, and you would avoid forcing TS users to pass both the generic and the standard parameter.
- Our naming convention discourages using abbreviations or single letters for everything. There are severa places where this is wrong, so feel free to change it if you find one of those cases.
- I guess the final type should be optional. I suggested below  using an inline type, but feel free to change it to a `Partial<Record<PropNames, string>>` if you find it easier to read.

```suggestion
export function dynamicPropsMixin<PropNames extends string>(
  propNames: PropNames[]
): ExtendedVue<Vue, unknown, unknown, unknown, { [Key in PropNames]?: string }> {
  return Vue.extend({
    props: propNames.reduce(
      (props, propName) => ({ ...props, [propName]: { type: String } }),
      {} as Record<PropNames, PropValidator<string>>
    )
  });
}
```